File: api/src/main/java/io/github/bindglam/neko/content/ContentType.java
package io.github.bindglam.neko.content;

import io.github.bindglam.neko.registry.Registries;
import org.jetbrains.annotations.NotNull;
import org.spongepowered.configurate.ConfigurationNode;

public interface ContentType<T extends Content> {
    @NotNull String id();

    @NotNull Class<T> clazz();

    boolean load(@NotNull Registries registries, @NotNull ConfigurationNode config);
}

File: core/src/main/kotlin/io/github/bindglam/neko/content/ContentType.kt
package io.github.bindglam.neko.content

import io.github.bindglam.neko.registry.Registries
import io.github.bindglam.neko.registry.WritableRegistry
import org.bukkit.configuration.ConfigurationSection

interface ContentType<T, E : WritableRegistry.RegistryEntry<T>> {
    val name: String

    fun load(registries: Registries, config: ConfigurationSection): LoadResult

    class LoadResult private constructor(val errorMsg: String?) {
        companion object {
            fun success() = LoadResult(null)
            fun failure(errorMsg: String) = LoadResult(errorMsg)
        }

        val isSuccess: Boolean
            get() = errorMsg == null
        val isFailure: Boolean
            get() = errorMsg != null
    }
}